contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,114 @@
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            ++num_rows;
+            if (num_rows > 20) {
+                eosio::check(false, "Cannot purchase tokens more than 20 times.");
+            }
+        }
+
+        // Calculate the number of seconds since sales start
+        uint32_t now_since_epoch = eosio::current_time_point().sec_since_epoch();
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        
+        vestingtoken::vesting_settings dates = launch_sales_dates_singleton.get();
+        uint32_t sales_start_date_since_epoch = dates.sales_start_date.sec_since_epoch();
+        
+        uint32_t allocated_after_sales_start_seconds = 0;
+        if (now_since_epoch > sales_start_date_since_epoch) {
+            allocated_after_sales_start_seconds = now_since_epoch - sales_start_date_since_epoch;
+        }
+
+        vesting_table.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.total_allocated = amount;
+            row.tokens_claimed = asset(0, amount.symbol);
+            row.seconds_since_sales_start = allocated_after_sales_start_seconds;
+            row.vesting_category_type = category;
+        });
+    }
+
+    void vestingToken::withdraw(eosio::name holder) {
+        require_auth(holder);
+
+        // Get the vesting allocations
+        vesting_allocations vesting_table(get_self(), holder.value);
+        uint32_t now_since_epoch = eosio::time_point_sec(eosio::current_time_point()).sec_since_epoch();
+
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        vestingtoken::vesting_settings dates = launch_sales_dates_singleton.get();
+        uint32_t launch_date_since_epoch = dates.launch_date.sec_since_epoch();
+        
+        for (auto iter = vesting_table.begin(); iter != vesting_table.end(); ++iter) {
+            const vested_allocation& vesting_allocation = *iter;
+
+            vesting_category category = vesting_allocation.vesting_category_type;
+
+            uint32_t vesting_start_since_epoch = launch_date_since_epoch + vesting_allocation.seconds_since_sales_start + category.start_delay_seconds;
+            uint32_t cliff_end_since_epoch = vesting_start_since_epoch + category.cliff_period_seconds;
+            
+            // Calculate the vesting end time
+            uint32_t vesting_end_time_since_epoch = vesting_start_since_epoch + category.vesting_period_seconds;
+
+            // Check if vesting period after cliff has started
+            eosio::check(now_since_epoch >= cliff_end_since_epoch, "Vesting period has not started");
+
+            // Calculate the total claimable amount
+            uint64_t claimable = 0;
```

Comment:
this should be `int64_t` not `uint64_t`
https://docs.antelope.io/cdt/latest/reference/Classes/structeosio_1_1asset/#variable-amount